pkg/ccl/backupccl/restore_job.go

```diff
@@ -355,31 +419,54 @@ func restore(
 		// to progCh.
 		for progress := range progCh {
 			mu.Lock()
-			var progDetails backuppb.RestoreProgress
-			if err := pbtypes.UnmarshalAny(&progress.ProgressDetails, &progDetails); err != nil {
-				log.Errorf(ctx, "unable to unmarshal restore progress details: %+v", err)
-			}
+			if execCtx.ExecCfg().Settings.Version.IsActive(restoreCtx, clusterversion.V23_1Start) {
+				var progDetails backuppb.RestoreProgress
+				if err := pbtypes.UnmarshalAny(&progress.ProgressDetails, &progDetails); err != nil {
+					log.Errorf(ctx, "unable to unmarshal restore progress details: %+v", err)
+				}
 
-			mu.res.Add(progDetails.Summary)
-			idx := progDetails.ProgressIdx
+				mu.res.Add(progDetails.Summary)
+				_, err = checkpointingSpanFrontier.Forward(progDetails.DataSpan, hlc.Timestamp{WallTime: 1})
+				if err != nil {
+					log.Errorf(ctx, "unable to forward timestamp: %+v", err)
+				}
 
-			// Assert that we're actually marking the correct span done. See #23977.
-			if !importSpans[progDetails.ProgressIdx].Span.Key.Equal(progDetails.DataSpan.Key) {
-				mu.Unlock()
-				return errors.Newf("request %d for span %v does not match import span for same idx: %v",
-					idx, progDetails.DataSpan, importSpans[idx],
-				)
-			}
-			mu.requestsCompleted[idx] = true
-			for j := mu.highWaterMark + 1; j < len(mu.requestsCompleted) && mu.requestsCompleted[j]; j++ {
-				mu.highWaterMark = j
+				mu.spanFrontierSlice = append(mu.spanFrontierSlice,
```

Comment:
ah, i think I mislead you. recall that one of the reasons we're using a spanFrontier is  to store less stuff in the job record. As you've currently implemented, we're storing each completed restore span entry, instead of the set of continuous completed spans.

`mu` should hold a spanFrontier. We shouldn't need a separate in memory `spanFrontierSlice`. Then when it's time to flush progress to disk, you call something like [this.](https://github.com/cockroachdb/cockroach/blob/47a03f07de10dbe012798d15a18905703e524c13/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go#L415)